contracts/schemes/JoinAndQuit.sol

```diff
@@ -0,0 +1,244 @@
+        bytes32 proposalId = votingMachine.propose(2, voteParams, msg.sender, address(avatar));
+        address member = _member;
+        if (member == address(0)) {
+            member = msg.sender;
+        }
+
+        Proposal memory proposal = Proposal({
+            accepted: false,
+            proposedMember: member,
+            funding : _feeAmount,
+            funder : msg.sender
+        });
+        proposals[proposalId] = proposal;
+
+        emit JoinInProposal(
+            address(avatar),
+            proposalId,
+            _descriptionHash,
+            member,
+            _feeAmount
+        );
+
+        proposalsInfo[address(votingMachine)][proposalId] = ProposalInfo({
+            blockNumber:block.number,
+            avatar:avatar
+        });
+        return proposalId;
+    }
+
+    /**
+    * @dev RedeemReputation reward for proposal
+    * @param _proposalId the ID of the voting in the voting machine
+    * @return reputation the redeemed reputation.
+    */
+    function redeemReputation(bytes32 _proposalId) public returns(uint256 reputation) {
+
+        Proposal memory _proposal = proposals[_proposalId];
+        Proposal storage proposal = proposals[_proposalId];
+        //set proposal proposedMember to zero to prevent reentrancy attack.
+        proposal.proposedMember = address(0);
+        require(proposal.accepted == true, " proposal not accepted");
+        require(
+        Controller(
+        avatar.owner()).mintReputation(memberReputation, _proposal.proposedMember));
+        proposal.proposedMember = _proposal.proposedMember;
+        emit RedeemReputation(address(avatar), _proposalId, _proposal.proposedMember, memberReputation);
+    }
+
+    /**
+    * @dev donate donate to the dao. no repution is minted
+    */
+    function donate(uint256 _donation) public {
+        address(fundingToken).safeTransferFrom(msg.sender, address(avatar), _donation);
+        fundings[msg.sender] = fundings[msg.sender].add(_donation);
+        totalDonation = totalDonation.add(_donation);
+        checkFundedBeforeDeadLine();
+        emit Donation(address(avatar), _donation);
+    }
+
+    /**
+    * @dev rageQuit quit from the dao.
+    * can be done on any time
+    * REFUND = USER_DONATION * CURRENT_DAO_BALANCE / TOTAL_DONATIONS
+    * @return refund the refund amount
+    */
+    function rageQuit() public returns(uint256 refund) {
+        require(fundings[msg.sender] > 0, "no fund to RageQuit");
+        uint256 userDonation = fundings[msg.sender];
+        fundings[msg.sender] = 0;
+        refund = userDonation.mul(fundingToken.balanceOf(address(avatar))).div(totalDonation);
+        totalDonation = totalDonation.sub(userDonation);
+        require(
+        Controller(
+        avatar.owner()).externalTokenTransfer(fundingToken, msg.sender, refund));
+        emit RageQuit(address(avatar), refund);
+    }
+
+    /**
+    * @dev checkFundedBeforeDeadLine check if funding goal reached.
```

Comment:
This does a bit more though than just check, though - it sets the db value
Call it `setFundingGoalReachedFlag` (it's ugly but at least it says what is does)